### ISSUE TITLE
Warn about "type(x) is C" or support it?

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#7260]
Mypy doesn't allow things like `type(x) is int` to be used to narrow down types, since this doesn't cover any subtypes of `int`. Maybe we should generate an error and suggest using `isinstance(...)` instead. Since this could be noisy, perhaps we should only do this if `x` is a union type.

Alternatively, we could support this as a valid type check, but treat it as an incomplete type check. This code would therefore still generate an error (but only one error):

```py
def f(x: Union[int, str]) -> None:
    if type(x) is int:
        return x + 1  # Should be okay? (currently generates an error)
    else:
        return x + 'foo'  # Still an error, since 'x' could be a subtype of 'int'
```

Also, if the target type `X` is final, using `type(x) is X` instead of `isinstance()` would be technically fine, though it's still stylistically suspect.

See #7258 for an example where this caused confusion.

**Comment by ilevkivskyi** [comment]
I think I am fine with supporting this. This is a bad style but we probably should not be too opinionated about style in cases where we can do safe reasoning about the code. I think we can indeed show only error in the second branch, and not show any errors at all in case of a final class.

**Comment by AlexWaygood** [comment]
Using `type(C) is int` for type narrowing [is now documented](https://mypy.readthedocs.io/en/latest/type_narrowing.html#type-narrowing-expressions) as being supported, and the example [now works as expected](https://mypy-play.net/?mypy=latest&python=3.10&gist=c1f0d8d18c07240ce1c47ab654385f8a) on mypy 0.942, so I'm closing this. (Feel free to reopen if I'm mistaken!)

```python
def f(x: int | str) -> None:
    if type(x) is int:
        x + 1  # No error on mypy 0.942
    else:
        x + 'foo'  # error: Unsupported operand types for + ("int" and "str")  # note: Left operand is of type "Union[int, str]"
```
